CIPs/cip-0032.md

```diff
@@ -0,0 +1,120 @@
+
+Rather than introducing a slashing policy which punishes poor performance, we could add additional reward to validators for great AS performance. We haven't ruled this option out and will likely institute it once we have more granular metrics (see [Using completion rate](#using-completion-rate)). Today, many validators put a good amount of effort into maintaining a highly available Attestation Service. Slashing based on poor performance ensures that everyone is putting in this same level of effort to grow our network for the same compensation.
+
+In order to slash effectively, we will need an on-chain measure of AS availability. We're researching ways to provide this measurement in a manner that ensures validators are correctly rewarded when their AS is healthy and correctly punished when their AS is down. Until we have a large number of users, it's difficult to isolate a signal we can be confident in as a community. Therefore, we're starting with a conservative approach of measuring a minimum completion threshold over a period of time.
+
+## Specification
+
+- `MinAttestationSlashablePeriod` is the number of consecutive blocks in which an elected Attestation Service is required to complete at least one attestation.
+- `AttestationSlasherPenalty` is the penalty amount that is slashed in a slashing event.
+- `AttestationSlasherReward` is the reward amount that a slasher is granted.
+- `MinAttestationSlashablePeriod`, `AttestationSlasherPenalty`, and `AttestationSlasherReward` are easily configurable via governance proposals.
+
+We will introduce a new `AttestationSlasher` contract which allows slashing of stake if a completed attestation hasn't been observed in the past `MinAttestationSlashablePeriod` blocks. To achieve this, we will upgrade the `Attestations` contract to register the selection and completion events in the new `AttestationSlasher` contract. The `AttestationSlasher` will contain the following state to track the latest block number that an attestation was requested/completed on:
+
+```js
+// Tracks the last time an attestation was requested from an issuer (AS node)
+mapping(address => uint256) public lastAttestationRequested;
+// Tracks the last time an attestation was completed by an issuer
+mapping(address => uint256) public lastAttestationCompleted;
+// Tracks the last time an issuer was last slashed
+mapping(address => uint256) public lastSlashed;
+```
+
+A slashing event may occur when the slashing initiator calls `slash()` with the slashable address. Before slashing, the slashing contract checks the following:
+
+1. The validator has been elected continuously for the past `MinAttestationSlashablePeriod` blocks.
+2. The issuer was selected for at least one attestation in the past `MinAttestationSlashablePeriod` blocks.
```

Comment:
The `MinAttestationSlashablePeriod` is the variable we would use to change the "difficulty" of the min-bar. The number of required attestations is the numerator and the `MinAttestationSlashablePeriod` is the denominator. We can adjust either to change the difficulty. I'm proposing holding the numerator constant since it also simplifies the state management in the contract.

Reply:
@YazzyYaz the initial values are proposed below. Are you looking for something like this?
> 1. The validator has been elected continuously for the past `MinAttestationSlashablePeriod` (190,080) blocks.

Reply:
Sorry, I just meant going from `one` to `1` for any values assigned to variables in a spec.